### PR TITLE
Add and switch to new types for revocation

### PR DESCRIPTION
**Pull request by canndrew**
Add the `UInt48`, `RevocationKey`, `CommitmentPubKey`, `CommitmentSeed` and `RevocationSet` types.

A `RevocationKey` is a secret key that can be used to revoke a transaction. It's shared in the `revoke_and_ack` message and stored in the user's wallet in a `RevocationSet`. A `CommitmentPubKey` is the public key of a `RevocationKey`, also known as a "per commitment point". A `CommitmentSeed` is the master revocation key for a side of a channel and can be used to derive all the revocation keys for that side of that channel.

A `RevocationSet` stores a set of revocation keys in a compact form specified in BOLT #3 which allows earlier keys to be derived from later keys.

### CODE REVIEW

**Comment by joemphilips** [comment]
Thanks, wow this looks significant improvement.
Give me a day, I want to understand what is going on...

File: tests/DotNetLightning.Core.Tests/RevocationSetTests.fs [hunk 1]
```diff
@@ -0,0 +1,281 @@
+module RevocationSetTests
+
+open NBitcoin
+open Expecto
+open DotNetLightning.Utils
+open DotNetLightning.Crypto
+
+let Unwrap<'T, 'E>(result: FSharp.Core.Result<'T, 'E>) : 'T =
```
Comment (joemphilips) [inline tests/DotNetLightning.Core.Tests/RevocationSetTests.fs:8]: nit: there is a `Result.deref` in ResultUtils to do this.

File: src/DotNetLightning.Core/Utils/UInt48.fs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        UInt64 = 1UL
+    }
+
+    static member MaxValue: UInt48 = {
+        UInt64 = UInt48.BitMask
+    }
+
+    static member FromUInt64(x: uint64): UInt48 =
+        if x > UInt48.BitMask then
+            failwith <| sprintf "value %i out of range for UInt48" x
```
Comment (joemphilips) [inline src/DotNetLightning.Core/Utils/UInt48.fs:25]: nit: raising `ArgumentOutOfRangeException` is better.

File: src/DotNetLightning.Core/Utils/NBitcoinExtensions.fs [hunk 3]
```diff
@@ -35,3 +37,24 @@ module NBitcoinExtensions =
 
         member this.GetTxId() =
             this.GetGlobalTransaction().GetTxId()
+
+    type Key with
+        static member BytesLength: int = 32
+
+        member this.ToByteArray(): array<byte> =
```
Comment (joemphilips) [inline src/DotNetLightning.Core/Utils/NBitcoinExtensions.fs:44]: Why you need this? `Key` already has `ToBytes()` method.

File: src/DotNetLightning.Core/Utils/NBitcoinExtensions.fs [hunk 4]
```diff
@@ -35,3 +37,24 @@ module NBitcoinExtensions =
...
+            let mutable bytes: array<byte> = Array.zeroCreate Key.BytesLength
+            use bytesStream = new MemoryStream(bytes)
+            let stream = NBitcoin.BitcoinStream(bytesStream, true)
+            this.ReadWrite stream
+            bytes
+
+    type PubKey with
+        static member BytesLength: int = 33
+
+        member this.ToByteArray(): array<byte> =
```
Comment (joemphilips) [inline src/DotNetLightning.Core/Utils/NBitcoinExtensions.fs:54]: Same as above.

**Comment by joemphilips** [comment]
There are few nits but overall it seems good.

**Comment by canndrew** [comment]
Thanks. I've fixed the nits (I didn't know about `ToBytes` or `deref`).

**Pull request merged**